File: src/cpp_pkg/src/test_node.cpp
#include "rclcpp/rclcpp.hpp"

class ROS2Node : public rclcpp::Node
{
public:
    ROS2Node() : Node("cpp_test")
    {
        RCLCPP_INFO(this->get_logger(), "Hello CPP Node");
        timer_ = this->create_wall_timer(std::chrono::seconds(1),
                                         std::bind(&ROS2Node::timerCallback, this));
    }

private:
    void timerCallback()
    {
        counter_++;
        RCLCPP_INFO(this->get_logger(), "Hello CPP Node %d", counter_);
    }

    rclcpp::TimerBase::SharedPtr timer_ ;
    int counter_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ROS2Node>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/py_pkg/py_pkg/test_node.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

class ROS2Node(Node):
    
    def __init__(self):
        super().__init__("py_test")
        self.counter = 0
        self.get_logger().info("Hello ROS2")
        self.create_timer(0.5, self.timer_callback)

    def timer_callback(self):
        self.counter += 1
        self.get_logger().info("Hello ROS2 !! " + str(self.counter))

def main(args=None):
    rclpy.init(args=args)
    node = ROS2Node()
    rclpy.spin(node)
    rclpy.shutdown

if __name__ == "__main__":
    main()


# --- Minimal example ---
#   def main(args=None):
#  rclpy.init(args=args)
#   node = Node("py_test")
#   node.get_logger().info("Hello ROS2")
#   rclpy.spin(node)
#   rclpy.shutdown